dgisim/src/dices.py

```diff
@@ -162,6 +175,205 @@ def just_satisfy(self, requirement: AbstractDices) -> bool:
+                filler_omni_count = 0
+            else:
+                filler_element_count = supply[filler_element]
+                filler_omni_count = need[Element.OMNI] - filler_element_count
+
+            result_dct[filler_element] += filler_element_count
+            result_dct[Element.OMNI] += filler_omni_count
+
+            need[Element.OMNI] -= filler_element_count + filler_omni_count
+            supply[filler_element] -= filler_element_count
+            supply[Element.OMNI] -= filler_omni_count
+
+            assert need[Element.OMNI] == 0, "need element omni not 0"
+            assert supply[filler_element] >= 0, "supply <0"
+            assert supply[Element.OMNI] >= 0, "supply <0"
+
+        # 2 step - fill the Elements needs and ANY (files ANY parallely)
+
+        global ELEMENTS_BY_DECREASING_GLOBAL_PRIORITY
+
+        reverse_order_need_list: List[Element] = list(
+            sorted(ELEMENTS_BY_DECREASING_GLOBAL_PRIORITY,
+                   key=lambda el: get_precedence(el)
+                   ))[::-1]  # + [Element.ANY]
+
+        for el in reverse_order_need_list:
+            if need[el] <= supply[el]:  # if element can fill itself
+                number_of_el = need[el]
+                # print(number_of_el)
+
+                result_dct[el] += number_of_el
+                need[el] -= number_of_el
+                supply[el] -= number_of_el
+                check_asserts()
+
+                if supply[el] > 0 and need[Element.ANY] > 0:
+                    number_of_el = min(supply[el], need[Element.ANY])
+                    print(number_of_el)
+
+                    result_dct[el] += number_of_el
+                    need[Element.ANY] -= number_of_el
+                    supply[el] -= number_of_el
+                    check_asserts()
+            else:  # if OMNI dices needs
+                number_of_el = supply[el]
+
+                result_dct[el] += number_of_el
+                need[el] -= number_of_el
+                supply[el] -= number_of_el
+
+                check_asserts()
+
+                if need[el] <= supply[Element.OMNI]:
+                    number_of_omni = need[el]
+
+                    result_dct[Element.OMNI] += number_of_omni
+                    need[el] -= number_of_omni
+                    supply[Element.OMNI] -= number_of_omni
+
+                    check_asserts()
+                else:
+                    return None
+
+        # 3 process ANY need with OMNI
+
+        if need[Element.ANY] > 0:
+            if supply[Element.OMNI] >= need[Element.ANY]:
+                number_of_spent_omni = need[Element.ANY]
+
+                result_dct[Element.OMNI] += number_of_spent_omni
+                need[Element.ANY] -= number_of_spent_omni
+                supply[Element.OMNI] -= number_of_spent_omni
+
+                check_asserts()
+            else:
+                return None
+
+        assert set(need.values()) == {0}
+        return ActualDices(dices=result_dct)
```

Comment:
The last part is ok, but in this project, most of the time, there is only requirement for a single kind of element (excluding OMNI and ANY), thus what I might do instead is:

- loop over pure elements needed, and fill with corresponding element or OMNI
- create a list of type `list[tuple[precedence, element]]` and heapify it. ([documentation of heap](https://docs.python.org/3/library/heapq.html)) (the list contains OMNI but has lowest priority for sure)
- make use of `heapq.heappop()` to fill ANY in precedence order.

Reply:
Sorry, don't understand enough. What is difference between it and my algorithm? I see using heapq and generalization of last part (filling ANY with OMNI), is it all?

Reply:
The main difference is it is slightly faster under most use cases of this simulator where the requests including ANY is likely to include a maximum of one kind of element as accompany.

You sorted the entire list of elements _O(nlogn)_ to likely use only the first two elements to fill up the ANY requirement, which is kind of an overkill just to fill ANY. (where to fill the other elements, you don't need  to sort the list) If you use heapq to extract the first two elements, then it should be faster than sorting the entire list.

BTW, you may want to check if number of dices you hold > dices required in the beginning, if not return None immediately. It can help accelerate the algorithm when the player is running low of dices.

Reply:
done this optimisation. Also added check on total supply >= total need at the beginning.